rcl_action/src/rcl_action/action_client.c

```diff
@@ -0,0 +1,723 @@
+            action_client->impl = impl;
+            RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Action client initialized");
+            return RCL_RET_OK;
+          }
+          RCUTILS_LOG_DEBUG_NAMED(
+            ROS_PACKAGE_NAME, "Finalizing action feedback subscription");
+          critical_ret = rcl_subscription_fini(&impl->feedback_subscription, node);
+          if (RCL_RET_OK != critical_ret) {
+            RCL_SET_ERROR_MSG("cleanup after error failed, "
+                              "system left inconsistent\n");
+          }
+          RCUTILS_LOG_DEBUG_NAMED(
+            ROS_PACKAGE_NAME, "Action feedback subscription finalized");
+        }
+        RCUTILS_LOG_DEBUG_NAMED(
+          ROS_PACKAGE_NAME, "Finalizing action result client");
+        critical_ret = rcl_client_fini(&impl->result_client, node);
+        if (RCL_RET_OK != critical_ret) {
+          RCL_SET_ERROR_MSG("cleanup after error failed, "
+                            "system left inconsistent\n");
+
+        }
+        RCUTILS_LOG_DEBUG_NAMED(
+          ROS_PACKAGE_NAME, "Action result client finalized");
+      }
+      RCUTILS_LOG_DEBUG_NAMED(
+        ROS_PACKAGE_NAME, "Finalizing action cancel client");
+      critical_ret = rcl_client_fini(&impl->cancel_client, node);
+      if (RCL_RET_OK != critical_ret) {
+        RCL_SET_ERROR_MSG("cleanup after error failed, "
+                          "system left inconsistent\n");
+      }
+      RCUTILS_LOG_DEBUG_NAMED(
+        ROS_PACKAGE_NAME, "Action cancel client finalized");
+    }
+    RCUTILS_LOG_DEBUG_NAMED(
+      ROS_PACKAGE_NAME, "Finalizing action goal client");
+    critical_ret = rcl_client_fini(&impl->goal_client, node);
+    if (RCL_RET_OK != critical_ret) {
+      RCL_SET_ERROR_MSG("cleanup after error failed, "
+                        "system left inconsistent\n");
+    }
+    RCUTILS_LOG_DEBUG_NAMED(
+      ROS_PACKAGE_NAME, "Action goal client finalized");
+  }
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Failed to initialize action client");
+  allocator->deallocate(impl->action_name, allocator->state);
+  allocator->deallocate(impl, allocator->state);
+  return ret;
+}
+
+rcl_ret_t
+rcl_action_client_fini(rcl_action_client_t * action_client, rcl_node_t * node)
+{
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Finalizing action client");
+  if (!rcl_action_client_is_valid(action_client)) {
+    return RCL_RET_ACTION_CLIENT_INVALID;
+  }
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_NODE_INVALID;
+  }
+  rcl_ret_t ret = RCL_RET_OK;
+  if (action_client->impl) {
+    // TODO(hidmic): ideally we should rollback to a valid state if any
+    //               finalization failed, but it seems that's currently
+    //               not possible.
+    rcl_action_client_impl_t * impl = action_client->impl;
+    if (rcl_client_is_valid(&impl->goal_client)) {
+      ret = rcl_client_fini(&impl->goal_client, node);
+      if (RCL_RET_OK != ret) {
+        return RCL_RET_ERROR;
```

Comment:
Recommend returning the original error code here, same below

Reply:
So, I tried to be consistent with `rcl_action` documentation instead of propagating internal errors (that are usually hard to understand because of the lack of context). If we propagate, documentation needs an update.